lib/src/importer/node_package.dart

```diff
@@ -0,0 +1,398 @@
+      var parent = p.dirname(entry);
+
+      // prevent infinite recursion
+      if (entry == parent) return null;
+
+      var rootLength = isWindows ? 1 : 0;
+
+      if (Uri.directory(parent).pathSegments.length == rootLength) return null;
+
+      return recurseUpFrom(parent);
+    }
+
+    return recurseUpFrom(baseDirectory);
+  }
+
+  /// Returns a file path specified by the `sass` or `style` values in a package
+  /// manifest, or an `index` file relative to the package root.
+  String? _resolvePackageRootValues(
+      String packageRoot, Map<String, dynamic> packageManifest) {
+    if (packageManifest['sass'] case String sassValue) {
+      if (validExtensions.contains(p.extension(sassValue))) {
+        return p.url.join(packageRoot, sassValue);
+      }
+    }
+
+    if (packageManifest['style'] case String styleValue) {
+      if (validExtensions.contains(p.extension(styleValue))) {
+        return p.url.join(packageRoot, styleValue);
+      }
+    }
+
+    var result = resolveImportPath(p.url.join(packageRoot, 'index'));
+    if (result != null) return result;
+    return null;
+  }
+
+  /// Returns a file path specified by a `subpath` in the `exports` section of
+  /// package.json.
+  ///
+  /// `packageName` is used for error reporting.
+  String? _resolvePackageExports(String packageRoot, String? subpath,
+      Map<String, dynamic> packageManifest, String packageName) {
+    if (packageManifest['exports'] == null) return null;
+    var exports = packageManifest['exports'] as Object;
+    var subpathVariants = _exportsToCheck(subpath);
+    if (_nodePackageExportsResolve(
+            packageRoot, subpathVariants, exports, subpath, packageName)
+        case String path) {
+      return path;
+    }
+
+    if (subpath != null && p.extension(subpath).isNotEmpty) return null;
+
+    var subpathIndexVariants = _exportsToCheck(subpath, addIndex: true);
+    if (_nodePackageExportsResolve(
+            packageRoot, subpathIndexVariants, exports, subpath, packageName)
+        case String path) {
+      return path;
+    }
+
+    return null;
+  }
+
+  /// Returns the path to one subpath variants, resolved in the `exports` of a
+  /// package manifest.
+  ///
+  /// Throws an error if multiple `subpathVariants` match, and null if none
+  /// match.
+  ///
+  /// Implementation of `PACKAGE_EXPORTS_RESOLVE` from the [Resolution Algorithm
+  /// Specification](https://nodejs.org/api/esm.html#resolution-algorithm-specification).
+  String? _nodePackageExportsResolve(
+      String packageRoot,
+      List<String?> subpathVariants,
+      Object exports,
+      String? subpath,
+      String packageName) {
+    if (exports is Map<String, dynamic>) {
+      if (exports.keys.any((key) => key.startsWith('.')) &&
+          exports.keys.any((key) => !key.startsWith('.'))) {
+        throw 'Invalid Package Configuration';
+      }
+    }
+    String? processVariant(String? subpath) {
```

Comment:
It's confusing to shadow the outer `subpath` parameter with this one. Consider renaming it.